build_tools/scripts/download_file.py

```diff
@@ -37,44 +43,66 @@ def parse_arguments():
   parser.add_argument("--unpack",
                       action='store_true',
                       default=False,
-                      help="Unpack the downloaded file if it's an archive.")
+                      help="Unpack the downloaded file if it's an archive")
+  parser.add_argument("--max-tries",
+                      metavar="<max-tries>",
+                      type=int,
+                      default=DEFAULT_MAX_TRIES,
+                      help="Number of tries before giving up")
   return parser.parse_args()
 
 
-def main(args):
-  output_dir = os.path.dirname(args.output)
-
-  if not os.path.isdir(output_dir):
-    os.makedirs(output_dir)
-
+def download_and_extract(source_url: str, output: str, unpack: bool):
   # Open the URL and get the file-like streaming object.
-  with urllib.request.urlopen(args.source_url) as response:
+  with urllib.request.urlopen(source_url) as response:
     if response.status != 200:
       raise RuntimeError(
           f"Failed to download file with status {response.status} {response.msg}"
       )
 
-    if args.unpack:
-      if args.source_url.endswith(".tar.gz"):
+    if unpack:
+      if source_url.endswith(".tar.gz"):
         # Open tar.gz in the streaming mode.
         with tarfile.open(fileobj=response, mode="r|*") as tar_file:
-          if os.path.exists(args.output):
-            shutil.rmtree(args.output)
-          os.makedirs(args.output)
-          tar_file.extractall(args.output)
+          if os.path.exists(output):
+            shutil.rmtree(output)
+          os.makedirs(output)
+          tar_file.extractall(output)
         return
-      elif args.source_url.endswith(".gz"):
+      elif source_url.endswith(".gz"):
         # Open gzip from a file-like object, which will be in the streaming mode.
         with gzip.open(filename=response, mode="rb") as input_file:
-          with open(args.output, "wb") as output_file:
+          with open(output, "wb") as output_file:
             shutil.copyfileobj(input_file, output_file)
         return
 
     # Fallback to download the file only.
-    with open(args.output, "wb") as output_file:
+    with open(output, "wb") as output_file:
       # Streaming copy.
       shutil.copyfileobj(response, output_file)
 
 
+def main(args):
+  output_dir = os.path.dirname(args.output)
+
+  if not os.path.isdir(output_dir):
+    os.makedirs(output_dir)
+
+  remaining_tries = args.max_tries
+  while remaining_tries > 0:
+    try:
+      download_and_extract(args.source_url, args.output, args.unpack)
+      break
+    except (ConnectionResetError, ConnectionRefusedError,
+            urllib.error.URLError):
+      remaining_tries -= 1
+      if remaining_tries == 0:
+        raise
+      else:
+        logging.warning(f"Connection error, remaining {remaining_tries} tries",
+                        exc_info=True)
+        time.sleep(RETRY_COOLDOWN_TIME)
```

Comment:
We're not at a scale where this necessarily matters, but it's generally best practice to use an exponential backoff (even better, with a little bit of randomness to avoid stampedes, but I think that doesn't matter much here)